src/main/kotlin/org/rust/ide/refactoring/changeSignature/impl.kt

```diff
@@ -0,0 +1,293 @@
+        config
+    ) { factory.createWhitespace(" ") }
+}
+
+private fun changeParameters(factory: RsPsiFactory, function: RsFunction, config: RsChangeFunctionSignatureConfig) {
+    val parameters = function.valueParameterList ?: return
+
+    importParameterTypes(config.parameters, function)
+    changeParametersNameAndType(factory, parameters.valueParameterList, config.parameters)
+    if (!config.parameterSetOrOrderChanged()) return
+
+    val parametersCopy = parameters.copy() as RsValueParameterList
+    fixParametersOrder(
+        factory,
+        parameters,
+        parametersCopy,
+        parametersCopy.selfParameter,
+        parametersCopy.valueParameterList,
+        config
+    ) { factory.createValueParameter(it.patText, it.typeReference, reference = false) }
+}
+
+private fun renameParameter(factory: RsPsiFactory, parameter: RsValueParameter, newName: String) {
+    val identifier = factory.createIdentifier(newName)
+
+    val binding = parameter.pat?.findBinding()
+    if (binding != null) {
+        val usages = RenameUtil.findUsages(binding, newName, false, false, emptyMap())
+        for (info in usages) {
+            RenameUtil.rename(info, newName)
+        }
+    }
+
+    binding?.identifier?.replace(identifier)
+}
+
+private fun fixParametersOrder(
+    factory: RsPsiFactory,
+    /** [RsValueArgumentList] or [RsValueParameterList] */
+    parameters: RsElement,
+    parametersCopy: RsElement,
+    parameterSelf: RsElement?,
+    parametersList: List<RsElement>,
+    config: RsChangeFunctionSignatureConfig,
+    createNewParameter: (Parameter) -> PsiElement
+) {
+    if (parametersList.size != config.originalParameters.size) return
+    val descriptors = config.parameters
+
+    // remove old parameters
+    val lparen = parameters.firstChild
+    val rparen = parameters.lastChild
+    if (lparen.nextSibling != rparen) {
+        parameters.deleteChildRange(lparen.nextSibling, rparen.prevSibling)
+    }
+    if (descriptors.isEmpty() && parameterSelf == null) return
+
+    // collect parameters psi in right order
+    val isMultiline = parametersCopy.textContains('\n')
+    val selfGroup = parameterSelf?.collectSurroundingWhiteSpaceAndComments()
+    val groupsWithoutSelf = descriptors.map { descriptor ->
+        if (descriptor.index == ParameterInfo.NEW_PARAMETER) {
+            val newline = if (isMultiline) factory.createNewline() else null
+            val psi = createNewParameter(descriptor)
+            listOfNotNull(newline, psi)
+        } else {
+            val psi = parametersList[descriptor.index]
+            psi.collectSurroundingWhiteSpaceAndComments()
+        }
+    }
+    val groups = listOfNotNull(selfGroup) + groupsWithoutSelf
+
+    // add parameters one by one
+    val rspace = parametersCopy.lastChild.prevSibling as? PsiWhiteSpace
+    val isSingleParameter = descriptors.size == 1 && descriptors.size < config.originalParameters.size
```

Comment:
```suggestion
    val isSingleParameter = groups.size == 1 && descriptors.size < config.originalParameters.size
```

Looks like I also forget about `self` parameter here. Also please add test for removing parameter in multiline method like:
```rust
impl Foo {
    fn foo(
        &self,
        a: i32,
        b: i32,
    ) {}
}
```